### ISSUE TITLE
Inserting new ActiveRecord attempts to set primary key to NULL

### ISSUE THREAD

**Issue by NilSet** [yiisoft/yii2#7374]
When inserting a new ActiveRecord corresponding to a database table consisting only of one autoincremening primary key column, it attemps to set the primary key to NULL. This causes issues in postgresql. It should instead insert a row of DEFAULT VALUES.

**Comment by samdark** [comment]
You mean key and nothing more? Is there a valid use case for it?

**Comment by samdark** [comment]
Anyway, Yii should not behave like that even in weird cases.

**Comment by cebe** [comment]
> it attemps to set the primary key to NULL

Can you elaborate what you mean with that?

**Comment by NilSet** [comment]
The following should be able to reproduce it.

Given a postgresql table defined by 
```sql
CREATE TABLE contact
(
  id serial NOT NULL,
  CONSTRAINT contact_pkey PRIMARY KEY (id)
);
```
and the corresponding Gii generated model,
the following php code
```php    
$contract = new \common\models\Contact();
$contract->save();
```
generates this query
```sql
INSERT INTO "contact" ("id") VALUES (NULL)
```

As for the use case, it is a foreign key in a number of places. I can work around this issue by giving the table an extra column and assigning the attribute null before saving.

**Comment by Faryshta** [comment]
what is the expected query?

**Comment by lynicidn** [comment]
i think yes, how ar know what is next pk needed?

**Comment by lynicidn** [comment]
http://stackoverflow.com/questions/7718585/how-to-set-auto-increment-primary-key-in-postgresql

```
CREATE TABLE epictable
(
    mytable_key    serial primary key,
);
```

**Comment by NilSet** [comment]
@Faryshta Generally attributes that have not been assigned any new data are not added to the insert query. This has been allowing the database to manage its own autoincrementing keys on all my tables. I would expect saving a model with no assigned attributes to generate `INSERT INTO "contact" DEFAULT VALUES`

**Comment by Faryshta** [comment]
thats a different behaviour from other db engines. `INSERT INTO "contact" ("id") VALUES (NULL)` would work perfectly on mysql and oracle AFAIK. Maybe an special case of 'save' would be required for posgresql

**Comment by NilSet** [comment]
@Faryshta Unfortunately in postgresql it errors out on violating the not-null constraint.

I have found that `INSERT INTO "contact" ("id") VALUES (DEFAULT)` also would work.

**Comment by nineinchnick** [comment]
This is triggered in `ActiveRecord::internalInsert()` when `ActiveRecord::getDirtyAttributes()` does not return anything. To avoid an `INSERT` without any column, primary keys with their defaults are used.

There isn't one way to build such `INSERT` statement that would work in each db, I've checked at least [pgsql](http://www.postgresql.org/docs/9.3/static/sql-insert.html) (doesn't allow `NULL`), [mysql](http://dev.mysql.com/doc/refman/5.6/en/insert.html) (no `DEFAULT VALUES` instead of `VALUES`) and [sqlite](http://sqlite.org/lang_insert.html) (no `DEAULT` instead of value).

To avoid adding conditions for db names in AR class, the default values for PKs should be read correctly, which are not. This makes this issue relate to #7843.

**Comment by GlyphDataServices** [comment]
Any fix for this yet?

**Comment by cebe** [comment]
This should be fixed in version 2.0.5, if you have an issue with a later version, please open a new ticket and describe the problem in detail including a way to reproduce it if possible.

**Comment by GlyphDataServices** [comment]
I had 2.0.5 but updating to 2.0.10 fixed it. Thanks!
